src/helper/sampler.hpp

```diff
@@ -0,0 +1,102 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog
+// (https://www.datadoghq.com/). Copyright 2022 Datadog, Inc.
+
+#pragma once
+
+#include <atomic>
+#include <cmath>
+#include <cstdint>
+#include <iostream>
+#include <mutex>
+#include <optional>
+
+namespace dds {
+static const double min_rate = 0.0001;
+class sampler {
+public:
+    sampler(double sample_rate)
+    {
+        // NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+        if (sample_rate <= 0) {
+            sample_rate = 0;
+        } else if (sample_rate > 1) {
+            sample_rate = 1;
+        } else if (sample_rate < min_rate) {
+            sample_rate = min_rate;
+        }
+
+        group_size_ = static_cast<unsigned>(round(1 / sample_rate));
+        // NOLINTEND(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+        tokens_per_group_ =
+            static_cast<unsigned>(round(sample_rate * group_size_));
+
+        reset_tokens();
+    }
+    class scope {
+    public:
+        explicit scope(std::atomic<bool> &concurrent) : concurrent_(&concurrent)
+        {
+            *concurrent_ = true;
+        }
+
+        scope(const scope &) = delete;
+        scope &operator=(const scope &) = delete;
+        scope(scope &&oth) noexcept
+        {
+            concurrent_ = oth.concurrent_;
+            oth.concurrent_ = nullptr;
+        }
+        scope &operator=(scope &&oth)
+        {
+            concurrent_ = oth.concurrent_;
+            oth.concurrent_ = nullptr;
+
+            return *this;
+        }
+
+        ~scope()
+        {
+            if (concurrent_ != nullptr) {
+                *concurrent_ = false;
+            }
+        }
+
+    protected:
+        std::atomic<bool> *concurrent_;
+    };
+
+    std::optional<scope> get()
+    {
+        const std::lock_guard<std::mutex> lock_guard(mtx_);
+
+        if (group_size_ == 1) {
+            if (!concurrent_) {
+                return {scope{concurrent_}};
+            }
+            return std::nullopt;
+        }
+
+        if (request_++ == group_size_) {
+            reset_tokens();
+            request_ = 1;
+        }
+        if (tokens_available_ > 0 && !concurrent_) {
+            tokens_available_--;
+            return {scope{concurrent_}};
+        }
+        return std::nullopt;
+    }
+
+protected:
+    void reset_tokens() { tokens_available_ = tokens_per_group_; }
+    unsigned request_{0};
+    std::atomic<bool> concurrent_{false};
+    std::mutex mtx_;
+    unsigned tokens_per_group_;
+    unsigned group_size_{1};
+    unsigned tokens_available_;
+};
```

Comment:
This seems all unnecessarily complication. It seems to me that all you need to do if to have a counter that you update atomically (it can even be with relaxed memory ordering) and just test if `floor(sample_rate * i) != floor(sample_rate * (i+1))`. Example:

```mathematica
oSample = Function[{r, i}, Floor[r*i] != Floor[r*(i + 1)]];
Select[Range[1, 20], doSample[.3, #] &]
Select[Range[1, 20], doSample[.75, #] &]
```

```
{3, 6, 9, 13, 16, 19}
{1, 2, 3, 5, 6, 7, 9, 10, 11, 13, 14, 15, 17, 18, 19}
```

Reply:
I went with this implementation. As I said, it blew my mind. Thanks for it